mantidproject/mantid
# MaterialBuilder exposed to python

The main goal was to expose `MaterialBuilder` to python. In the process there was a refactor of the underlying `Material` class and how it stores information. This was caused by a need to reproduce various factors used by PDF measurements.

**To test:**

This should be most verified by review and seeing that all of the tests still pass. A script that produces (in standard units) most of "gemsheet" is:

``` python
import mantid
from mantid.kernel import MaterialBuilder
import math

def printSummary(material):
    chem = material.chemicalFormula()
    print 'Sym   AMass    Num   c    b bar  b     Sigma s  Sigma a'
    totalAtoms = 0.
    for amount in chem[1]:
        totalAtoms += amount

    for (atom, amount) in zip(chem[0], chem[1]):
        neutron = atom.neutron()
        bbar = neutron['coh_scatt_length_real'] # according to gemsheet
        b = neutron['tot_scatt_length']         # according to gemsheet
        print '%3s%9.4f%6.2f%6.2f%7.4f%7.4f%8.4f%8.4f' % (atom.symbol, atom.mass,
                                                          amount,
                                                          amount/totalAtoms,
                                                          bbar,
                                                          b,
                                                          neutron['tot_scatt_xs'], # Sigma s
                                                          neutron['abs_xs']) # Sigma a

    print "(sum c*b^2)     %.4f" % mat.totalScatterLengthSqrd() # E12
    print "(sum c bbar)^2  %.4f" % (mat.cohScatterLengthReal()*mat.cohScatterLengthReal()) # E14

builder = MaterialBuilder()
mat = builder.setFormula('Ba Ti O3').build()
print '#################### Ba Ti O3'
printSummary(mat)

builder = MaterialBuilder()
mat = builder.setFormula('Al2 O3').build()
print '#################### Al2 O3'
printSummary(mat)
```

It does not generate the Faber-Ziman factors, but the information is now easily created in python.

_There is no associated issue_



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
